mesonbuild/envconfig.py

```diff
@@ -237,6 +237,12 @@ def get_java_home(self) -> T.Optional[Path]:
         value = T.cast('T.Optional[str]', self.properties.get('java_home'))
         return Path(value) if value else None
 
+    def get_bindgen_clang_args(self) -> T.List[str]:
+        value = mesonlib.listify(self.properties.get('bindgen_clang_arguments', []))
+        if not all(isinstance(v, str) for v in value):
+            raise EnvironmentException('bindgen_clang_arguments must be a string or an array of strings')
+        return T.cast('T.List[str]', value)
```

Comment:
Shouldn't this be stringlistify, which incidentally returns `T.List[str]` for you already?

Reply:
Hmm, bit disappointing... maybe we can just make stringlistify/typeslistify accept a `prefix=` argument that will be prepended to "(List) item must be" in error messages?

Reply:
AFAICT basically every use of typeslistify/stringlistify should go away when we finish the `typed_kwargs` conversion, because they're basically all things like: `x = stringlistyf(extract_as_list(kwargs, 'foo'))`. `typeslistify` also doesn't flatten, so we're still stuck with `mesonlib.stringlistify(mesonlib.listify(self.properties.get('bindgen_clang_arguments', [])))`

Reply:
Do we even normally expect flattening in a  machine file, where it's not standard to have things that may be strings, may be arrays, and may be combined using various operators some of which append and some of which extend?

Reply:
Plus, we're still having to catch it to convert the type from a generic MesonException to EnvironmentException.

Reply:
(But actually I do think the error message from typeslistify is a strict improvement modulo the prefix :p)

Reply:
> Do we even normally expect flattening in a machine file

Yes. Or we don't do any validation and die in a fire, or we just assert a bunch of stuff with no messages :D

But for things that may be arrays listify() is in fact called

Reply:
I opened https://github.com/mesonbuild/meson/pull/11912 for the `prefix` thing. Since this PR is super important for mesa and the merge window is about to close, I'd *really* like to get this in today. If you want me to switch to using `stringlistify`, I'd rather do that as part of #11912 after this lands.